torch/nn/attention/templated_attention.py

```diff
@@ -0,0 +1,165 @@
+    return scores @ value
+
+
+@sdpa.py_impl(DispatchKey.CompositeExplicitAutograd)
+def sdpa_dense(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    score_mod: Callable,
+    *other_buffers: torch.Tensor,
+):
+    return math_attention(query, key, value, score_mod, *other_buffers).contiguous()
+    # TODO we want to convert his to a call to sdpa and not math_attention
+
+
+# TODO For now lets disable but we need to figure this out
+sdpa.py_impl(DispatchKey.Autograd)(autograd_not_implemented(sdpa, deferred_error=True))
+
+
+def trace_sdpa(
+    proxy_mode,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    score_mod: Callable,
+    *other_buffers: torch.Tensor,
+):
+    if score_mod is None:
+        with proxy_mode:
+            return F.scaled_dot_product_attention(query, key, value)
+
+    with disable_proxy_modes_tracing():
+        example_out = F.scaled_dot_product_attention(query, key, value)
+    example_vals = [torch.zeros((), dtype=query.dtype)] + [
+        torch.zeros((), dtype=torch.int) for _ in range(4)
+    ]
+    score_graph = make_fx(score_mod)(*example_vals, *other_buffers)
+    proxy_mode.tracer.root.register_module("sdpa_score", score_graph)
+    node_args = (query, key, value, score_graph, *other_buffers)
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, node_args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function", sdpa, proxy_args, {}, name="sdpa_impl"
+    )
+    return track_tensor_tree(
+        example_out, out_proxy, constant=None, tracer=proxy_mode.tracer
+    )
+
+
+@sdpa.py_impl(ProxyTorchDispatchMode)
+def sdpa_proxy_torch_dispatch_mode(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    score_mod: Callable,
+    *other_buffers: torch.Tensor,
+):
+    mode = _get_current_dispatch_mode()
+    assert mode is not None, "Mode should always be enabled for python fallback key"
+    with _pop_mode_temporarily() as mode:
+        if mode.enable_tracing:
+            return trace_sdpa(mode, query, key, value, score_mod, *other_buffers)
+        else:
+            return sdpa(query, key, value, score_mod, *other_buffers)
+
+
+@sdpa.py_functionalize_impl
+def sdpa_functionalize(
+    ctx,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    score_mod: Callable,
+    *other_buffers: torch.Tensor,
+):
+    query_unwrapped = ctx.unwrap_tensors(query)
+    key_unwrapped = ctx.unwrap_tensors(key)
+    value_unwrapped = ctx.unwrap_tensors(value)
+    # score_mod_unwrapped = ctx.(score_mod)
+    other_buffers_unwrapped = ctx.unwrap_tensors(other_buffers)
+    # type(list(score_mod.graph.nodes)[1].meta['example_value'])
+    with ctx.redispatch_to_next() as m:
+        functional_score_mod = ctx.functionalize(score_mod)
+        pre_dispatch = hasattr(ctx, "mode") and ctx.mode.pre_dispatch
+        example_vals = [torch.zeros((), dtype=query.dtype)] + [
```

Comment:
You might need to move the `torch.zero()` creation to be outside of the `ctx.redispatch_to_next()` context.

I think that wil cause the extra tensors you create to be wrapped up into `FunctionalTensors`, but the idea is that the tensors you pass to the inner `sdpa()` call should all be "unwrapped"